beacon_node/eth1/src/web3_fetcher.rs

```diff
@@ -0,0 +1,336 @@
+use bls::{PublicKeyBytes, SignatureBytes};
+use ethabi::{decode, ParamType, Token};
+use parking_lot::RwLock;
+use std::collections::BTreeMap;
+use std::fs;
+use std::marker::Send;
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::prelude::*;
+use types::DepositData;
+use web3::contract::{Contract, Options};
+use web3::futures::Future;
+use web3::transports::WebSocket;
+use web3::types::FilterBuilder;
+use web3::types::*;
+use web3::Web3;
+
+use crate::error::{Error, Result};
+use crate::types::Eth1DataFetcher;
+
+/// Wrapper around web3 api.
+/// Transport hardcoded to ws since its needed for subscribing to logs.
+#[derive(Clone, Debug)]
+pub struct Web3DataFetcher {
+    event_loop: Arc<web3::transports::EventLoopHandle>,
+    /// Websocket transport object. Needed for logs subscription.
+    web3: Arc<web3::api::Web3<web3::transports::ws::WebSocket>>,
+    /// Deposit Contract
+    contract: Contract<web3::transports::ws::WebSocket>,
+}
+
+impl Web3DataFetcher {
+    /// Create a new Web3 object.
+    pub fn new(
+        endpoint: &str,
+        deposit_contract_addr: &str,
+        deposit_contract_abi: PathBuf,
+    ) -> Result<Self> {
+        let (event_loop, transport) = WebSocket::new(endpoint)?;
+        let abi = fs::read(deposit_contract_abi).map_err(|_| Error::InvalidParam)?;
+        let web3 = Web3::new(transport);
+        let contract = Contract::from_json(
+            web3.eth(),
+            deposit_contract_addr
+                .parse()
+                .map_err(|_| Error::InvalidParam)?,
+            &abi,
+        )?;
+        Ok(Web3DataFetcher {
+            event_loop: Arc::new(event_loop),
+            web3: Arc::new(web3),
+            contract: contract,
+        })
+    }
+}
+
+impl Eth1DataFetcher for Web3DataFetcher {
+    /// Get block_number of current block.
+    fn get_current_block_number(&self) -> Box<dyn Future<Item = U256, Error = Error> + Send> {
+        Box::new(
+            self.web3
+                .eth()
+                .block_number()
+                .map_err(|e| {
+                    println!("Error getting block number");
+                    Error::Web3Error(e)
+                })
+                .timeout(Duration::from_secs(10))
+                .map_err(|_| {
+                    println!("Timed out getting block_number");
+                    Error::Timeout
```

Comment:
I believe this error isn't always a timeout, it can also be an error in the request.

Reply:
This will always be a `Timeout` error. The error in the request is handled in the previous `map_err`